pkg/components/pubsub/pluggable_test.go

```diff
@@ -0,0 +1,293 @@
+	pullCalled      atomic.Int64
+	pullErr         error
+}
+
+//nolint:nosnakecase
+func (s *server) PullMessages(svc proto.PubSub_PullMessagesServer) error {
+	s.pullCalled.Add(1)
+
+	if s.onAckReceived != nil {
+		go func() {
+			for {
+				msg, err := svc.Recv()
+				if err != nil {
+					return
+				}
+				s.onAckReceived(msg)
+			}
+		}()
+	}
+	if s.pullChan != nil {
+		for msg := range s.pullChan {
+			if err := svc.Send(msg); err != nil {
+				return err
+			}
+		}
+	}
+
+	return s.pullErr
+}
+
+func (s *server) Init(_ context.Context, req *proto.PubSubInitRequest) (*proto.PubSubInitResponse, error) {
+	s.initCalled.Add(1)
+	if s.onInitCalled != nil {
+		s.onInitCalled(req)
+	}
+	return &proto.PubSubInitResponse{}, s.initErr
+}
+
+func (s *server) Features(context.Context, *proto.FeaturesRequest) (*proto.FeaturesResponse, error) {
+	s.featuresCalled.Add(1)
+	return &proto.FeaturesResponse{}, s.featuresErr
+}
+
+func (s *server) Publish(_ context.Context, req *proto.PublishRequest) (*proto.PublishResponse, error) {
+	s.publishCalled.Add(1)
+	if s.onPublishCalled != nil {
+		s.onPublishCalled(req)
+	}
+	return &proto.PublishResponse{}, s.publishErr
+}
+
+func (s *server) Ping(context.Context, *proto.PingRequest) (*proto.PingResponse, error) {
+	s.pingCalled.Add(1)
+	return &proto.PingResponse{}, s.pingErr
+}
+
+func TestPubSubPluggableCalls(t *testing.T) {
+	getPubSub := testingGrpc.TestServerFor(testLogger, func(s *grpc.Server, svc *server) {
+		proto.RegisterPubSubServer(s, svc)
+	}, func(cci grpc.ClientConnInterface) *grpcPubSub {
+		client := proto.NewPubSubClient(cci)
+		pubsub := fromConnector(testLogger, pluggable.NewGRPCConnector(components.Pluggable{}, proto.NewPubSubClient))
+		pubsub.Client = client
+		return pubsub
+	})
+
+	if runtime.GOOS != "windows" {
+		t.Run("test init should populate features and call grpc init", func(t *testing.T) {
+			const (
+				fakeName          = "name"
+				fakeType          = "type"
+				fakeVersion       = "v1"
+				fakeComponentName = "component"
+				fakeSocketFolder  = "/tmp"
+			)
+
+			uniqueID := guuid.New().String()
+			socket := fmt.Sprintf("%s/%s.sock", fakeSocketFolder, uniqueID)
+			defer os.Remove(socket)
+
+			connector := pluggable.NewGRPCConnectorWithFactory(func(string) string {
+				return socket
+			}, proto.NewPubSubClient)
+			defer connector.Close()
+
+			listener, err := net.Listen("unix", socket)
+			require.NoError(t, err)
+			defer listener.Close()
+			s := grpc.NewServer()
+			srv := &server{}
+			proto.RegisterPubSubServer(s, srv)
+			go func() {
+				if serveErr := s.Serve(listener); serveErr != nil {
+					testLogger.Debugf("Server exited with error: %v", serveErr)
+				}
+			}()
+
+			ps := fromConnector(testLogger, connector)
+			err = ps.Init(pubsub.Metadata{
+				Base: contribMetadata.Base{},
+			})
+
+			require.NoError(t, err)
+			assert.Equal(t, int64(1), srv.featuresCalled.Load())
+			assert.Equal(t, int64(1), srv.initCalled.Load())
+		})
+	}
```

Comment:
Should we fatal/abort/log anything if trying to execute this from windows? That's something I was thinking about pluggable components and windows, beyond this test: I think we are not giving Windows user enough feedback as that pluggable components won't run in windows. Enabling it should fail. Running this test should fail.